### PR TITLE
Cross-platform compatible xattr user namespace

### PR DESCRIPTION
**Pull request by freqlabs**
### Motivation and Context

ZFS on Linux originally implemented xattr namespaces in a way that is incompatible with other operating systems.  On illumos, xattrs do not have namespaces.  Every xattr name is visible.  FreeBSD has two universally defined namespaces: EXTATTR_NAMESPACE_USER and EXTATTR_NAMESPACE_SYSTEM.  The system namespace is used for protected FreeBSD-specific attributes such as MAC labels and pNFS state.  These attributes have the namespace string "freebsd:system:" prefixed to the name in the encoding scheme used by ZFS.  The user namespace is used for general purpose user attributes and obeys normal access control mechanisms.  These attributes have no namespace string prefixed, so xattrs written on illumos are accessible in the user namespace on FreeBSD, and xattrs written to the user namespace on FreeBSD are accessible by the same name on illumos.

Linux has several xattr namespaces.  On Linux, ZFS encodes the namespace in the xattr name for every namespace, including the user namespace.  As a consequence, an xattr in the user namespace with the name "foo" is stored by ZFS with the name "user.foo" and therefore appears on FreeBSD and illumos to have the name "user.foo" rather than "foo".  Conversely, none of the xattrs written on FreeBSD or illumos are accessible on Linux unless the name
happens to be prefixed with one of the Linux xattr namespaces, in which case the namespace is stripped from the name.  This makes xattrs entirely incompatible between Linux and other platforms.

We want to make the encoding of user namespace xattrs compatible across platforms.  A crucial requirement of this compatibility is for xattrs from existing pools from FreeBSD and illumos to be accessible by the same names in the user namespace on Linux.  It is also necessary that existing pools with xattrs written by Linux retain access to those xattrs by the same names on Linux.  Making user namespace xattrs from Linux accessible by the correct names on other platforms is important. The handling of other namespaces is not required to be consistent.

### Description

Add a fallback mechanism for listing and getting xattrs to treat xattrs as being in the user namespace if they do not match a known prefix.

When setting user namespace xattrs, do not prefix the namespace to the name.  If the xattr is already present with the namespace prefix, remove it so only the non-prefixed version persists.  This ensures other platforms will be able to read the xattr with the correct name.

Do not allow setting or getting xattrs with a name that is prefixed with one of the namespace names used by ZFS on supported platforms.

Make xattr namespace compatibility dependent on a new feature.  New pools will use the compatible namespace encoding by default, and existing pools will continue using the old Linux-specific encoding until the feature is enabled.

Allow choosing between cross-platform compatability and legacy Linux compatibility with a per-dataset property.  This facilitates replication between hosts with different compatibility needs.

Add xattr_fallback property to skip fallbacks.

This property currently defaults to "on" so we do not miss xattrs in datasets from pools for which the encoding cannot be known or may even be a mix of both.  It can be turned "off" for a performance boost when it is known only the configured xattr_compat format is relevant.

Future work will investigate how we can change the default to "off" and automatically force it to "on" when we cannot know whether skipping the fallbacks is safe.

TODO:

- [x] If xattrs with the user namespace prefix are already present, they
  will not be automatically fixed.  Any existing xattrs must be
  manually rewritten on Linux for the name to be correct on FreeBSD.
  This also means that files may have a mix of incompatible and
  compatible names.  Other platforms could strip the Linux user
  namespace prefix from xattr names so they are presented correctly.

- [x] The newly written xattrs will no longer be visible on previous
  versions of ZFS on Linux.  This behavior needs to be made optional
  with a feature flag and possibly a per-dataset property.

- [x] There is no attempt to handle xattr names that clash with a namespace
  prefix.  If you write an xattr named "user.foo" to the user namespace
  on FreeBSD, the "user." prefix will be stripped on Linux.  This was
  partially the case already, except now the stripped name will also
  replace the prefixed name when updating the xattr.  Likewise, setting
  an xattr to the user namespace using a name with the prefix of
  another namespace may cause the xattr to be manipulated in the other
  namespace.  This is potentially a security issue.  Such names must be
  forbidden.

- [x] New tests should be added when the functionality is complete.

- [x] Documentation will be needed.

- [x] Performance optimizations should be investigated.

### How Has This Been Tested?

So far just existing ZTS tests and manually checking xattr naming/visibility on snaps sent between FreeBSD and Linux.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by freqlabs** [comment]
Ideally I'd like to be able to switch between the old and the new behavior with a per-dataset property. My intent with this would be to retain compatibility with older ZFS on Linux versions for replicated filesystems. Would this be possible? I'm not sure if the pool feature would prevent replication with older ZFS versions if it is not active on the specific dataset, or if the property would end up breaking replication because the other pool doesn't know about it.

**Comment by freqlabs** [comment]
I've not bothered storing the ABI change for the sake of this RFC, it would just clutter up the diff.

**Comment by freqlabs** [comment]
* Rebased
* Add feature@xattr_compat
* WIP xattr_compat property (needs tests, docs)

**Comment by freqlabs** [comment]
I've noticed that FreeBSD does not actually handle xattr=sa, it behaves as if it were xattr=on. I have gone ahead and implemented xattr=sa for FreeBSD but it is not included in this RFC as it does not directly relate to the proposed change to user namespace handling on Linux. It will be something I include in my test matrix.

**Comment by freqlabs** [comment]
* Rebased
* WIP xattr_compat for FreeBSD

**Comment by freqlabs** [comment]
* Rebased

**Comment by lundman** [comment]
We noticed that our xattr tests would fail, especially when going from xattr=sa to xattr=on, or v-v. This inspired us to take Linux xattr code and re-apply it;  https://github.com/openzfsonosx/openzfs/commit/a1c2e6c7d90a0ec499c5c9a30d9f1aa576ea45b4

The upshot of this is that macOS xattrs sources (albeit a different filename) is nearly identical to Linux (but better, as we take uio instead of that tmp struct! :) )
https://github.com/openzfsonosx/openzfs/blob/development/module/os/macos/zfs/zfs_vnops_osx_xattr.c

But, as for the xattr name-space, the macro `XATTR_USER_PREFIX` is still in there but not used. The name-space code can also be moved over, depending on what the consensus is?

**Comment by freqlabs** [comment]
* Rebased (with #12356)
* Squashed
* Added xattr_fallback boolean property to enable skipping fallback lookups as an optimization
* Added xattr_compat and xattr_fallback properties to general fs prop lists in tests
* Added xattr_compat feature to general feature list in tests

**Review by amotin — APPROVED** [review]

**Comment by freqlabs** [comment]
* Rebased so I can grab the ABI files from the CI

**Comment by freqlabs** [comment]
* Rebased

File: module/os/linux/zfs/zpl_xattr.c [hunk 1]
```diff
@@ -706,19 +704,33 @@ static int
...
 		return (-EOPNOTSUPP);
 
-	xattr_name = kmem_asprintf("%s%s", XATTR_USER_PREFIX, name);
-	error = zpl_xattr_get(ip, xattr_name, value, size);
-	kmem_strfree(xattr_name);
+	/*
+	 * Try to look up the name without the namespace prefix first for
+	 * compatibility with xattrs from other platforms.  If that fails,
+	 * try again with the namespace prefix if fallback is allowed.
+	 */
```
Comment (behlendorf) [inline module/os/linux/zfs/zpl_xattr.c:724]: It seems to me what we really want is a deterministic way to know if an xattr was written before or after the `xattr_compat` feature was enabled.  If it was written before then we need to invoke the fallback logic since it may or may not be prefixed.  However, if it was written after there's no ambiguity and thus no need for any fallback logic.

One way to do this would be to automatically upgrade all of xattrs attached to a znode when it's first accessed if 1) the `xattr_compat` feature is enabled and 2) and xattr upgraded bit isn't yet set.  There are available bits in the SA_ZPL_FLAGS system attribute we could reserve for this purpose.  All newly written xattrs could set this bit by default.   This would have the effect of gradually migrating all xattrs to the compatible format and the fallback logic would only be invoked once.  One nice this about this is then we shouldn't need a new fallback property.

> Allow choosing between cross-platform comparability and legacy Linux compatibility with a per-dataset property. This facilitates replication between hosts with different compatibility needs.

I believe what you're getting at here is send/recv compatibility since there currently is no one way to store a user xattr for all platforms.  This would allow a send stream to be generated which would be compatible with the xattr format used in older version of ZFS on Linux.

File: module/zcommon/zfeature_common.c [hunk 2]
```diff
@@ -598,6 +598,19 @@ zpool_feature_init(void)
 	zfeature_register(SPA_FEATURE_DRAID,
 	    "org.openzfs:draid", "draid", "Support for distributed spare RAID",
 	    ZFEATURE_FLAG_MOS, ZFEATURE_TYPE_BOOLEAN, NULL);
+
+
+	{
+	static const spa_feature_t xattr_compat_deps[] = {
+		SPA_FEATURE_EXTENSIBLE_DATASET,
```
Comment (behlendorf) [inline module/zcommon/zfeature_common.c:605]: As a little future insurance we should consider depending on `SPA_FEATURE_ENABLED_TXG`.  This way it'll always be possible to determine when the feature was enabled, and if we always write xattrs with the compat format it would be possible to determine what format they're in (even if we normally use a different mechanism for this).

File: module/os/linux/zfs/zpl_xattr.c [hunk 3]
```diff
@@ -706,19 +704,33 @@ static int
...
+	error = -ENODATA;
+	if (compat || fallback)
+		error = zpl_xattr_get(ip, name, value, size);
+	if (!compat || (fallback && error == -ENODATA)) {
+		char *xattr_name;
+		xattr_name = kmem_asprintf("%s%s", XATTR_USER_PREFIX, name);
```
Comment (behlendorf) [inline module/os/linux/zfs/zpl_xattr.c:730]: Since we're effectively redefining the on-disk format for user xattrs names with this change we can define what that should be.  Have you given any thought to making the prefixed "user." version the new format instead?  If all xattrs includes a namespace prefix the names would be consistent and less ambiguous.

File: module/os/linux/zfs/zpl_xattr.c [hunk 4]
```diff
@@ -729,19 +741,44 @@ __zpl_xattr_user_set(struct inode *ip, const char *name,
...
+	boolean_t compat = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_COMPAT);
+	boolean_t fallback = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_FALLBACK);
+	int error = 0;
 	/* xattr_resolve_name will do this for us if this is defined */
 #ifndef HAVE_XATTR_HANDLER_NAME
 	if (strcmp(name, "") == 0)
 		return (-EINVAL);
 #endif
+	if (ZFS_XA_NS_PREFIX_FORBIDDEN(name))
+		return (-EINVAL);
```
[the remark this answers is not in the document]
Comment (freqlabs) [reply module/os/linux/zfs/zpl_xattr.c:753]: The names come in here without the namespace prefix, and the name with the prefix stripped could still prefix match a namespace. For example, a nefarious user could attempt to write an xattr named "security.selinux" to the "user" namespace, which would then be interpreted as an xattr in the "security" namespace. These checks prevent that.

File: module/os/linux/zfs/zpl_xattr.c (hunk 1 above)
Comment (freqlabs) [reply module/os/linux/zfs/zpl_xattr.c:724]: A user xattr written before the xattr_compat feature is enabled *might* need the fallback, iff it was written on Linux. If it was written on any other platform, the fallback is not needed. We don't know in ZFS which platform things were written on a priori, so the xattr_fallback property allows for the fallback to be explicitly opted out of for performance when you know the xattrs written before xattr_compat did not come from Linux.

This does seem potentially useful still. One of my open questions is how to automatically skip the fallback logic in cases where we can assert it is safe to do so. We also have to know the xattr_compat property was never set to xattr_compat=linux, but if there are flags we can leverage that might do the trick.

Send/recv compatibility is the purpose of the xattr_compat property, yes.

File: module/os/linux/zfs/zpl_xattr.c (hunk 3 above)
Comment (freqlabs) [reply module/os/linux/zfs/zpl_xattr.c:730]: I have considered it, and while I do appreciate the approach for its cleanliness and internal consistency, I went in the direction of consistency with the existing implementations outside of Linux. There is a touch of aversion to the extra asprintfs required for the prefix as well, but I haven't actually attempted to measure any performance impact that might have.

File: module/os/linux/zfs/zpl_xattr.c (hunk 1 above)
Comment (behlendorf) [reply module/os/linux/zfs/zpl_xattr.c:724]: > One of my open questions is how to automatically skip the fallback logic in cases where we can assert it is safe to do so.

My feeling is it would be a nice simplification if we could also remove the need for the `xattr_compat` property. Then we could always write the data in the xattr_compat on-disk format and never need the fallback logic for upgraded znodes.  You'd just always pay the upgrade cost on first access after enabling the feature and then never again (except for read-only filesystems and snapshots).  We'd just need to find somewhere to store the bit indicating the znode's xattrs were upgraded. The SA_ZPL_FLAGS system attribute is one possible easy option.  We could probably also devine it based on birth times in the block pointers storing the xattrs (sa and directory), but I suspect we'd need to violate some layering to get that information so it may not be a great option.  If we thought about it some more I'm sure we could come up with other alternatives. 

The only catch I think is send/recv.  It might be possible to add a new `zfs send` flag which could be used to rewrite the xattr names in the send stream when its generated.  Or alternately, and perhaps much better, we could always generate the xattr_compat format and the `zstream` command could be taught to translate existing streams in to either format.  This way you could still generate a stream normally and then after the fact tailor it for target systems running older software.

File: module/os/linux/zfs/zpl_xattr.c (hunk 3 above)
Comment (behlendorf) [reply module/os/linux/zfs/zpl_xattr.c:730]: Well and the other argument for no-prefixes I can sympathize with is it does make the xattrs slightly smaller which means they're a little more likely to fit in the dnode SA space and not spill.  When that happens it does save a read I/O which is good for performance.

**Comment by freqlabs** [comment]
* Rebased
* Restored zpool-features.7 changes lost in rebase after man pages moved
* Add dependency on SPA_FEATURE_ENABLED_TXG
* Activate feature on FreeBSD when writing an xattr with xattr_compat=linux

File: module/os/linux/zfs/zpl_xattr.c (hunk 1 above)
Comment (freqlabs) [reply module/os/linux/zfs/zpl_xattr.c:724]: I'm not entirely on board with the idea of rewriting xattrs on znode access. That seems like it will potentially create unexpected load spikes that have no obvious explanation. We may potentially stat a very large number of files at once (such as for listdir), and that will create a lot of xattr rewrites. With xattr=sa there is still anIt's only once per file, but there's also the issue of read-only datasets such as snapshots. Samba exposes previous versions of files saved in snapshots using the .zfs dir. The snapshot data can't be rewritten, so we'll need all the same mechanisms for ensuring access to xattrs in those datasets anyway.

The properties allow users of both platforms to control the performance/compatibility tradeoffs they are willing to make. Without them, we leave cases where we cannot make safe assumptions about compatibility and must sacrifice performance. These cases are the very cross-platform migration situations we're trying to facilitate. Given an existing pool to be migrated, all xattrs were written before the feature was enabled, so we can't make any assumption in ZFS about the way the xattr names are stored. Upgrading the pool before the migration doesn't change the data already on disk.

* FreeBSD -> Linux
    The user knows this pool came from FreeBSD. We cannot determine this in ZFS, but the user can tell us by setting xattr_compat=all. The user also knows they have never written xattrs in this pool from Linux, so they don't need a fallback mechanism. We cannot determine this in ZFS, but the user can tell us by setting xattr_fallback=off, allowing us to avoid a performance regression on Linux.

* Linux -> FreeBSD
    The user knows this pool came from Linux. ZFS can't know this except by the xattr_compat property. The user also knows they have never written xattrs in this pool from another platform, so they are all in xattr_compat=linux format. ZFS can't know this except by xattr_fallback=off, allowing us to avoid a performance regression on FreeBSD.

The properties also allow us to avoid performance regression on Linux when upgrading existing pools that are only used on Linux. Setting xattr_compat=linux and xattr_fallback=off means the xattr format will remain unchanged and we will not need to pay the cost for compat fallbacks.

The above situations also apply when receiving datasets from another pool rather than importing a whole pool from another platform. The properties allow for separate datasets to have different formats, so for example one pool can receive streams from a mix of different systems for backup purposes, without requiring the user to remember to perform some extra translation step on their streams.

There are surely ways we could automatically ignore xattr_fallback=on and skip fallback when it can be guaranteed safe. I don't think we can avoid introducing both properties entirely without penalizing users switching between platforms. We have already experienced xattr operations being a bottleneck for Samba, which uses primarily xattrs to implement the MS-FSA layer. Making sure we have the options to avoid additional overhead in these operations is important.

**Comment by freqlabs** [comment]
* Rebased

File: module/os/linux/zfs/zpl_xattr.c [hunk 5]
```diff
@@ -729,19 +741,44 @@ __zpl_xattr_user_set(struct inode *ip, const char *name,
...
 	kmem_strfree(xattr_name);
 
+	if (!compat || error == -EEXIST)
+		return (error);
+	if (fallback && error == 0 && (flags & XATTR_REPLACE))
+		flags &= ~XATTR_REPLACE;
+	error = zpl_xattr_set(ip, name, value, size, flags);
+
+	dsl_dataset_t *ds = dmu_objset_ds(ITOZSB(ip)->z_os);
+	ds->ds_feature_activation[SPA_FEATURE_XATTR_COMPAT] = (void *)B_TRUE;
```
Comment (problame) [inline module/os/linux/zfs/zpl_xattr.c:780]: Judging from the current uses of `ds_feature_activation` in `master`, I think it must only be accessed from syncing context. Otherwise, the following two problems occur:

**1. Data Race** between reader (syncing context) and writer (the entity setting `ds_feature_activation[fi] = ...`.
**2. Too-late activation** of the feature, i.e., `dsl_dataset_sync` observes `ds_feature_activation[fi] == B_TRUE` in a later txg than the `zpl_xattr_set` changes persist to disk, which require the feature to be activated on read.

Could someone familiar with the zfeature code please confirm or correct my assessment?

(cc'ing @pcd1193182 because he refactored it a few years ago, and @ahrens as the original author of zfeature)

**Comment by freqlabs** [comment]
>     * Rebased
> 
>     * Restored zpool-features.7 changes lost in rebase after man pages moved
> 
>     * Add dependency on SPA_FEATURE_ENABLED_TXG
> 
>     * Activate feature on FreeBSD when writing an xattr with xattr_compat=linux

* Restored these changes that got lost somehow in the last push

File: module/os/linux/zfs/zpl_xattr.c (hunk 5 above)
Comment (freqlabs) [reply module/os/linux/zfs/zpl_xattr.c:780]: I think this is a good observation. I wonder if there is a way to activate the feature in the same txg? Or maybe activating it first and assuming success is the simplest option here...

**Comment by freqlabs** [comment]
* Fixed missing header on FreeBSD

**Comment by freqlabs** [comment]
* Removed feature@xattr_compat flag
* Removed xattr_fallback property
I'll fix the ABI files and squash after the CI gets a run at it.

File: module/os/freebsd/zfs/zfs_vnops_os.c [hunk 6]
```diff
@@ -5414,19 +5445,29 @@ zfs_getextattr(struct vop_getextattr_args *ap)
...
 		return (error);
 
 	error = ENOENT;
 	ZFS_ENTER(zfsvfs);
 	ZFS_VERIFY_ZP(zp)
 	rw_enter(&zp->z_xattr_lock, RW_READER);
-	if (zfsvfs->z_use_sa && zp->z_is_sa)
-		error = zfs_getextattr_sa(ap, attrname);
+
+	boolean_t compat = (zfsvfs->z_flags & ZSB_XATTR_COMPAT) != 0;
```
[the remark this answers is not in the document]
Comment (freqlabs) [reply module/os/freebsd/zfs/zfs_vnops_os.c:5468]: I wrote the platform code to follow the platform's conventions, which is !! on Linux and != 0 on FreeBSD. I'll gladly change these flags to booleans in a followup PR.

**Comment by freqlabs** [comment]
* Rebased
* Squashedd

**Review by behlendorf — APPROVED** [review]
Looks good, thanks for refreshing this.

**Comment by freqlabs** [comment]
I'll incorporate the updated ABI files after my local tests finish running.

**Comment by behlendorf** [comment]
You might want add a basic cross-compatibility test to verify that existing xattrs written in the `xattr_compat=linux` format are accessible when `xattr_compat=all`.  And even better that they do get rewritten when accessed.

**Comment by freqlabs** [comment]
Ok, on it. And I extracted the wrong archive of ABI files in that last update, oops.

**Comment by freqlabs** [comment]
I've added a test which caught a few edge cases on FreeBSD, which I've now fixed. We'll see how it runs in Linux in the CI.

**Comment by freqlabs** [comment]
Failing an edge case on Linux as well, so look forward to one more update.

**Comment by freqlabs** [comment]
I think that should be good for now. I would like to continue extending the test cases, but in the meantime what I've added has been useful enough already.

**Comment by behlendorf** [comment]
@freqlabs can you rebase this one last time, that should get us a clean build on all the builders.  @ahrens @lundman @problame if you could give this a final review I'd like to move forward with integrating this PR.

**Comment by freqlabs** [comment]
* Rebased

**Comment by freqlabs** [comment]
* Make xattr_compat=linux the default on Linux for now

**Comment by freqlabs** [comment]
* Rebase
* Make xattr_compat a tunable, zfs_xattr_compat

**Comment by behlendorf** [comment]
@freqlabs thanks for updating this so quickly.  You'll want to make sure to rebase after e2909fae8f2c3b15459fd4c0acc1999c6907864b to sort out the GitHub actions CI failure.  Sorry about that.

File: include/sys/fs/zfs.h [hunk 7]
```diff
@@ -1749,6 +1747,35 @@ enum zio_encrypt {
...
+#define	ZFS_XA_NS_PREFIX_MATCH(ns, name) \
+    (strncmp(name, ZFS_XA_NS_##ns##_PREFIX, ZFS_XA_NS_##ns##_PREFIX_LEN) == 0)
+
+#define	ZFS_XA_NS_PREFIX_FORBIDDEN(name) \
+    (ZFS_XA_NS_PREFIX_MATCH(FREEBSD, name) || \
+    ZFS_XA_NS_PREFIX_MATCH(LINUX_SECURITY, name) || \
+    ZFS_XA_NS_PREFIX_MATCH(LINUX_SYSTEM, name) || \
+    ZFS_XA_NS_PREFIX_MATCH(LINUX_TRUSTED, name) || \
+    ZFS_XA_NS_PREFIX_MATCH(LINUX_USER, name))
+/* END CSTYLED */
```
Comment (nabijaczleweli) [inline include/sys/fs/zfs.h:1778]: Why CSTYLED and not just
```c
#define	ZFS_XA_NS_PREFIX_MATCH(ns, name) \
	(strncmp(name, ZFS_XA_NS_##ns##_PREFIX, \
	    ZFS_XA_NS_##ns##_PREFIX_LEN) == 0)

#define	ZFS_XA_NS_PREFIX_FORBIDDEN(name) \
	(ZFS_XA_NS_PREFIX_MATCH(FREEBSD, name) || \
	    ZFS_XA_NS_PREFIX_MATCH(LINUX_SECURITY, name) || \
	    ZFS_XA_NS_PREFIX_MATCH(LINUX_SYSTEM, name) || \
	    ZFS_XA_NS_PREFIX_MATCH(LINUX_TRUSTED, name) || \
	    ZFS_XA_NS_PREFIX_MATCH(LINUX_USER, name))
```

i.e., cat -A:
```
#define^IZFS_XA_NS_PREFIX_MATCH(ns, name) \$
^I(strncmp(name, ZFS_XA_NS_##ns##_PREFIX, \$
^I    ZFS_XA_NS_##ns##_PREFIX_LEN) == 0)$
$
#define^IZFS_XA_NS_PREFIX_FORBIDDEN(name) \$
^I(ZFS_XA_NS_PREFIX_MATCH(FREEBSD, name) || \$
^I    ZFS_XA_NS_PREFIX_MATCH(LINUX_SECURITY, name) || \$
^I    ZFS_XA_NS_PREFIX_MATCH(LINUX_SYSTEM, name) || \$
^I    ZFS_XA_NS_PREFIX_MATCH(LINUX_TRUSTED, name) || \$
^I    ZFS_XA_NS_PREFIX_MATCH(LINUX_USER, name))$
```

File: man/man4/zfs.4 [hunk 8]
```diff
@@ -2129,6 +2129,30 @@ When enabled, the maximum number of pending allocations per top-level vdev
...
+.It Sy zfs_xattr_compat Ns = Ns 0 Ns | Ns 1 Pq int
+Control the naming scheme used when setting new xattrs in the user namespace.
+If
+.Sy 0
+(the default on Linux), user namespace xattr names are prefixed with the
+namespace, to be backwards compatible with previous versions of ZFS on Linux.
+If
+.Sy 1
+(the default on
+.Fx ),
```
Comment (nabijaczleweli) [inline man/man4/zfs.4:2141]: Wrong. `.Fx ) ,` or `.Pq the default on Fx ,`.

**Comment by freqlabs** [comment]
* Incorporated feedback

File: tests/zfs-tests/tests/functional/xattr/xattr_compat.ksh [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+log_assert "The zfs_xattr_compat tunable and fallback works as expected"
+log_onexit cleanup
+
+TESTFILE=$TESTDIR/testfile.$$
+TMPFILE=$TEST_BASE_DIR/tmpfile.$$
+OLD_XATTR_COMPAT=$(get_tunable XATTR_COMPAT)
+
+for x in sa dir; do
+	log_must zfs set xattr=$x $TESTPOOL/$TESTFS
+	log_must set_tunable32 XATTR_COMPAT 0
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/xattr/xattr_compat.ksh:61]: We could slightly increase the test coverage by writing the xattrs initially in the native platform format, then flipping back and forth for the subsequent tests steps.

File: module/os/linux/zfs/zpl_xattr.c [hunk 10]
```diff
@@ -706,19 +706,29 @@ static int
...
 	if (!(ITOZSB(ip)->z_flags & ZSB_XATTR))
 		return (-EOPNOTSUPP);
 
-	xattr_name = kmem_asprintf("%s%s", XATTR_USER_PREFIX, name);
-	error = zpl_xattr_get(ip, xattr_name, value, size);
-	kmem_strfree(xattr_name);
+	/*
+	 * Try to look up the name without the namespace prefix first for
+	 * compatibility with xattrs from other platforms.  If that fails,
+	 * try again with the namespace prefix.
```
Comment (behlendorf) [inline module/os/linux/zfs/zpl_xattr.c:723]: Since we're treating these both as equally valid formats, and I'd expect most pools not to move between platforms, why don't we check for the native format first to help minimize any potential performance impact.

**Review by behlendorf — APPROVED** [review]

**Comment by ahrens** [comment]
I don't have time to review the code thoroughly, but I discussed the design with @freqlabs and @behlendorf and I think it's great.  The explanation in the tunables manpage is super helpful!
